### PR TITLE
[server] Handle exceptions in do_POST

### LINKED ISSUE #2332: backslash in user name causes exception in ldap requests
If the user name contains backslash an exception is throuwn by the server.
We should escape special characters in the ldap searches.

INFO 2019-09-24 07:20] - 'user\name' logged in.
[ERROR 2019-09-24 07:20] - Filter error, please recheck your filter patterns.
[WARNING 2019-09-24 07:20] - {'desc': 'Bad search filter'}
Traceback (most recent call last):
  File "/codechecker/lib/python2.7/codechecker_server/server.py", line 448, in do_POST
    processor.process(iprot, oprot)
  File "/codechecker/lib/python2.7/codechecker_api/Authentication_v6/codeCheckerAuthentication.py", line 618, in process
    self._processMap[name](self, seqid, iprot, oprot)
  File "/codechecker/lib/python2.7/codechecker_api/Authentication_v6/codeCheckerAuthentication.py", line 663, in process_performLogin
    result.success = self._handler.performLogin(args.authMethod, args.authString)
  File "/codechecker/lib/python2.7/codechecker_common/profiler.py", line 77, in wrapper
    return release_wrapper(*args, **kwargs)
  File "/codechecker/lib/python2.7/codechecker_common/profiler.py", line 69, in release_wrapper
    res = function(*args, **kwargs)
  File "/codechecker/lib/python2.7/codechecker_server/api/authentication.py", line 85, in performLogin
    session = self.__manager.create_or_get_session(auth_string)
  File "/codechecker/lib/python2.7/codechecker_server/session_manager.py", line 552, in create_or_get_session
    validation = self.__handle_validation(auth_string)
  File "/codechecker/lib/python2.7/codechecker_server/session_manager.py", line 352, in __handle_validation
    validation = self.__try_auth_ldap(auth_string)
  File "/codechecker/lib/python2.7/codechecker_server/session_manager.py", line 455, in __try_auth_ldap
    if cc_ldap.auth_user(ldap_conf, username, password):
  File "/codechecker/lib/python2.7/codechecker_server/auth/cc_ldap.py", line 321, in auth_user
    account_scope)
  File "/codechecker/lib/python2.7/codechecker_server/auth/cc_ldap.py", line 149, in get_user_dn
    user_data = con.search_s(account_base_dn, scope, account_pattern)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 597, in search_s
    return self.search_ext_s(base,scope,filterstr,attrlist,attrsonly,None,None,timeout=self.timeout)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 590, in search_ext_s
    msgid = self.search_ext(base,scope,filterstr,attrlist,attrsonly,serverctrls,clientctrls,timeout,sizelimit)
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 586, in search_ext
    timeout,sizelimit,
  File "/usr/local/lib/python2.7/dist-packages/ldap/ldapobject.py", line 106, in _ldap_call
    result = func(*args,**kwargs)
FILTER_ERROR: {'desc': 'Bad search filter'}
pen.

**Desktop (please complete the following information)**
 - OS: [e.g. Linux]
 - Browser [e.g. Firefox, Chrome]

**Additional context**
Add any other context about the problem.

### PR DESCRIPTION
**Pull request by csordasmarton**
> Closes #2332 

Is it possible that a Python exception was raised in a Thrift API
function. We will catch this exception in the do_POST function and
we will try to send back a valid thrift message to the client. First
we try to call the `readMessageBegin` function on the input protocol
(iprot) but it will wait indefinitely because we have called this
function in the `processor.process(iprot, oprot)` function already
and no newer message can be found in the buffer.

To solve this problem we will create a new JSON protocol (`iprot`) by
using the value from the previous one and send a thrift exception
by using this new input protocol.

### CODE REVIEW

**Comment by csordasmarton** [comment]
Okay, I found the solution to this problem and I've rebased this commit.

One more thing. Do we want to handle LDAP exceptions in the server code or do we want to send it back to the user. For example the user will get the following message:
```
[ERROR 2019-09-25 13:25] - Internal server error: {u'info': 'Resource temporarily unavailable', 'errno': 11, 'desc': u'Bad search filter'}
```
If we handle these exceptions in the server the user will see that the authentication failed.

File: web/server/codechecker_server/auth/cc_ldap.py [hunk 1]
```diff
@@ -145,14 +145,17 @@ def get_user_dn(con,
...
-            # User found use the user DN from the first result.
-            user_dn = user_data[0][0]
-            LOG.debug("Found user: %s", user_dn)
-            return user_dn
+            if user_data:
+                # User found use the user DN from the first result.
+                user_dn = user_data[0][0]
+                LOG.debug("Found user: %s", user_dn)
+                return user_dn
+    except:
```
Comment (bruntib) [inline web/server/codechecker_server/auth/cc_ldap.py]: I still think we shouldn't hide the exception. At least we should print it to the "warning" level log. Same thing in `check_group_membership()`.

**Comment by csordasmarton** [comment]
@bruntib If you check the `ldap_error_handler` context manager, we will print the error messages there:
https://github.com/Ericsson/codechecker/blob/5fbdc80d14f494e18c0f2dcd649b110bf70a9328/web/server/codechecker_server/auth/cc_ldap.py#L119-L136

File: web/server/codechecker_server/server.py [hunk 2]
```diff
@@ -481,6 +481,9 @@ def do_POST(self):
             LOG.warning(str(exn))
             import traceback
             traceback.print_exc()
+
+            itrans = TTransport.TMemoryBuffer(itrans.cstringio_buf.getvalue())
```
Comment (gyorb) [inline web/server/codechecker_server/server.py]: Please check if there is data in `itrans.cstringio_buf.getvalue()` it is possible that the exception is thrown before the earlier processing of the thrift data.

**Comment by gyorb** [comment]
Please create a PR for the release v6.10.1 branch too

**Pull request merged**